src/main/kotlin/com/nikitanazarov/coroutinestacks/CoroutineStacksPanel.kt

```diff
@@ -31,34 +57,275 @@ class CoroutineStacksPanel(project: Project) : JBPanelWithEmptyText() {
+            val (node, level) = stack.pop()
+
+            if (node.value.stackTrace.isNotEmpty()) {
+                addCoroutineInfoBox(node.value, rows[level])
+            }
+
+            for (child in node.children) {
+                stack.push(child to level + 1)
+            }
+        }
+    }
+
+    private fun addCoroutineInfoBox(node: CoroutineStacksNode, box: Box) {
+        val headerText = CoroutineStacksBundle.message("number.of.coroutines", node.additionalData.size)
+        val stackFrames = mutableListOf<String>()
+        stackFrames.add(headerText)
+
+        stackFrames.addAll(node.stackTrace)
+
+        val coroutineListView = JBList<String>(stackFrames)
+
+        coroutineListView.cellRenderer = cellRenderer
+
+        val scrollPane = JBScrollPane(coroutineListView)
+
+        val border = BorderFactory.createLineBorder(JBColor.BLACK, Constants.borderWidth)
+        scrollPane.border = border
+
+        box.add(scrollPane)
+        box.add(Box.createHorizontalStrut(Constants.boxHorizontalStruct))
     }
+
+    private fun generateParallelStackTree(
+        tree: Tree<CoroutineStacksNode>,
+        rootValue: CoroutineStacksNode?,
+        positionOfStackFrame: Int
+    ) {
+        val dataToStackFrame = mutableMapOf<CoroutineInfoData, String>()
+        if (rootValue == null) {
+            return
+        }
+        for (data in rootValue.additionalData) {
+            if (data.stackTrace.size > positionOfStackFrame)
+                dataToStackFrame[data] = data.stackTrace[data.stackTrace.size -1 - positionOfStackFrame].toString()
+        }
+
+        val entries = dataToStackFrame.entries
+
+        val groupedByValue = entries.groupBy { it.value }
+
+        val groupedPositions = groupedByValue.map { (_, list) ->
+            list.map { it.key }
+        }
+
+        for (groupedPosition in groupedPositions) {
+            if (entries.size == 1 && rootValue.stackTrace.isNotEmpty()) {
+                rootValue.stackTrace.add(dataToStackFrame[groupedPosition[0]]!!)
+                generateParallelStackTree(tree, rootValue, positionOfStackFrame + 1)
+            } else if (entries.size > 1) {
+                val childValue = CoroutineStacksNode(stackTrace = mutableListOf(dataToStackFrame[groupedPosition[0]]!!), additionalData = groupedPosition)
+                println("childValue in recursion: $childValue")
+                println()
+                tree.insert(childValue, rootValue)
+                generateParallelStackTree(tree, childValue, positionOfStackFrame + 1)
+            }
+        }
+    }
+
+    data class CoroutineStacksNode(
+        val stackTrace: MutableList<String>,
+        val additionalData: List<CoroutineInfoData>
+    )
+    class TreeNode<T>(val value: T) {
+        val children: MutableList<TreeNode<T>> = mutableListOf()
+
+        fun addChild(child: TreeNode<T>) {
+            children.add(child)
+        }
+    }
+
+    class Tree<T> {
+        var root: TreeNode<T>? = null
```

Comment:
Instead of having a nullable root here, we could just require passing it to the constructor:
```
class Tree<T>(val root: TreeNode<T>)
```